src/robot_state_publisher.cpp

```diff
@@ -68,110 +73,311 @@ geometry_msgs::msg::TransformStamped kdlToTransform(const KDL::Frame & k)
+  segments_.clear();
+  segments_fixed_.clear();
+  addChildren(tree.getRootSegment());
+
+  auto msg = std::make_unique<std_msgs::msg::String>();
+  msg->data = urdf_xml;
+
   // Publish the robot description
-  description_pub_->publish(model_xml_);
+  description_pub_->publish(std::move(msg));
 }
 
 // add children to correct maps
 void RobotStatePublisher::addChildren(const KDL::SegmentMap::const_iterator segment)
 {
-  auto root = GetTreeElementSegment(segment->second).getName();
+  const std::string & root = GetTreeElementSegment(segment->second).getName();
 
-  auto children = GetTreeElementChildren(segment->second);
+  std::vector<KDL::SegmentMap::const_iterator> children = GetTreeElementChildren(segment->second);
   for (unsigned int i = 0; i < children.size(); i++) {
     const KDL::Segment & child = GetTreeElementSegment(children[i]->second);
     SegmentPair s(GetTreeElementSegment(children[i]->second), root, child.getName());
     if (child.getJoint().getType() == KDL::Joint::None) {
-      if (model_.getJoint(child.getJoint().getName()) &&
-        model_.getJoint(child.getJoint().getName())->type == urdf::Joint::FLOATING)
+      if (model_->getJoint(child.getJoint().getName()) &&
+        model_->getJoint(child.getJoint().getName())->type == urdf::Joint::FLOATING)
       {
-        fprintf(stderr,
-          "Floating joint. Not adding segment from %s to %s.\n",
+        RCLCPP_INFO(get_logger(),
+          "Floating joint. Not adding segment from %s to %s.",
           root.c_str(), child.getName().c_str());
       } else {
         segments_fixed_.insert(make_pair(child.getJoint().getName(), s));
-        fprintf(stderr, "Adding fixed segment from %s to %s\n", root.c_str(),
+        RCLCPP_DEBUG(get_logger(), "Adding fixed segment from %s to %s", root.c_str(),
           child.getName().c_str());
       }
     } else {
       segments_.insert(make_pair(child.getJoint().getName(), s));
-      fprintf(stderr, "Adding moving segment from %s to %s\n", root.c_str(),
+      RCLCPP_DEBUG(get_logger(), "Adding moving segment from %s to %s", root.c_str(),
         child.getName().c_str());
     }
     addChildren(children[i]);
   }
 }
 
-
 // publish moving transforms
 void RobotStatePublisher::publishTransforms(
   const std::map<std::string, double> & joint_positions,
-  const builtin_interfaces::msg::Time & time,
-  const std::string & tf_prefix)
+  const builtin_interfaces::msg::Time & time)
 {
-  // fprintf(stderr, "Publishing transforms for moving joints\n");
+  RCLCPP_DEBUG(get_logger(), "Publishing transforms for moving joints");
   std::vector<geometry_msgs::msg::TransformStamped> tf_transforms;
 
   // loop over all joints
-  for (auto jnt = joint_positions.begin(); jnt != joint_positions.end(); ++jnt) {
-    auto seg = segments_.find(jnt->first);
+  for (const std::pair<std::string, double> & jnt : joint_positions) {
+    std::map<std::string, SegmentPair>::iterator seg = segments_.find(jnt.first);
```

Comment:
Curious as to why you removed `auto` here

Reply:
I find `auto` hard to read.  I never know what type the variable is, so I don't know what methods I can call on it.  I prefer to be explicit.  Same goes for below.